java/ql/src/semmle/code/java/Annotation.qll

```diff
@@ -149,10 +267,54 @@ class AnnotationType extends Interface {
 
   /** Holds if this annotation type is annotated with the meta-annotation `@Inherited`. */
   predicate isInherited() {
-    exists(Annotation ann |
-      ann.getAnnotatedElement() = this and
-      ann.getType().hasQualifiedName("java.lang.annotation", "Inherited")
-    )
+    getADeclaredAnnotation().getType().hasQualifiedName("java.lang.annotation", "Inherited")
+  }
+
+  /** Holds if this annotation type is annotated with the meta-annotation `@Documented`. */
+  predicate isDocumented() {
+    getADeclaredAnnotation().getType().hasQualifiedName("java.lang.annotation", "Documented")
+  }
+
+  /**
+   * Gets the retention policy of this annotation type, that is, the name of one of the
+   * enum constants of `java.lang.annotation.RetentionPolicy`. If no explicit retention
+   * policy is specified the result is `CLASS`.
+   */
+  string getRetentionPolicy() {
+    if getADeclaredAnnotation() instanceof RetentionAnnotation
+    then result = getADeclaredAnnotation().(RetentionAnnotation).getRetentionPolicy()
+    else
+      // If not explicitly specified retention is CLASS
+      result = "CLASS"
+  }
+
+  /**
+   * Holds if the element type is a possible target for this annotation type.
+   * The `elementType` is the name of one of the `java.lang.annotation.ElementType`
+   * enum constants. If no explicit target is specified for this annotation type
+   * it is considered to be applicable to all elements.
+   */
+  // Note: Cannot use a predicate with string as result because annotation type without
+  // explicit @Target can be applied to all targets, requiring to hardcode element types here
+  bindingset[elementType]
```

Comment:
If there is a small set of possible values you can do something like `class AnnotationTargetType extends string { AnnotationTargetType() { this = ["CLASS", "METHOD", ...] } }` and drop the `bindingset`

Reply:
Yes that would be doable, but I wanted to avoid this (see "Note: .." comment a few lines above) because the set of possible target types would have to be adjusted then whenever a new constant is added to [`ElementType`](https://docs.oracle.com/en/java/javase/16/docs/api/java.base/java/lang/annotation/ElementType.html) (recents ones were `MODULE` and `RECORD_COMPONENT`).

Additionally it might cause confusing results when a Java database is not using the latest Java version. Consider a Java 8 project for which this predicate claims that `RECORD_COMPONENT` (added for Java 16) is a target.
But maybe that is not actually a big issue, especially since this predicate already relies on JDK-8231435 which was fixed only for Java 14.

I would prefer not using `bindingset` here as well.